### PR TITLE
Professional Development Feature

### PR DESCRIPTION
**Pull request by perryzjc**
### [Pivotal Tracker Link][tracker]

[tracker]: https://www.pivotaltracker.com/story/show/187375078

## What this PR does:

This pull request fixes|implements **implements** professional development feature for admins to create PD workshops and add teachers to each workshop

#### CURD for PD
image
image
image
image

#### CURD for PD Registrations
image
image
image

#### Who authored this PR?
1. Perry (Jingchao) Zhong (integrate frontend + backend, frontent mockup)
2. Michael Tao (integrate frontend + backend)
3. Jackson Xu (some backend implementation)

### How should this PR be tested?

* Is there a deploy we can view?
* What do the specs/features test?
1. Test Basic CURD for PD and PD registrations
2. Test UX workflow & value validations
* Are there edge cases to watch out for?

#### Are there any complications to deploying this?
1. Run `rails db:migrate`

### Checklist:

- [ ] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

### CODE REVIEW

**Comment by ArushC** [comment]
@perryzjc Overall this looks good. Everything works as intended. But there were two main things that I noticed. 

1. There seems to be an extreme lack of comments for the number of lines of code that were written. I don't think Michael really cares, and if it's too much work to write comments then don't put any. But some more comments would be helpful to include to understand what's going on at a **high level**, especially in some of the 100+ line code files that you added. 

2. You are missing some "sad path" tests according to the CodeCov warnings (see below). The sad path stuff obviously works based on the code that I see. But it shouldn't be too hard to get ChatGPT to right a few quick tests. 


Here are specific comments:

1) def set_pd_registration
    @pd_registration = PdRegistration.find(params[:id])
  rescue ActiveRecord::RecordNotFound
    redirect_to professional_development_path, alert: "Registration not found." ---> **THIS PATH IS MISSING TESTING** (probably needs to be tested using RSpec, since frontend won't go to an invalid path)

2) def set_professional_development
    @professional_development = ProfessionalDevelopment.find_by(id: params[:professional_development_id])
    unless @professional_development
      redirect_to professional_developments_path, alert: "Professional Development not found." --> **THIS PATH IS MISSING TESTING** (also needs to be tested using RSpec, since frontend won't go to an invalid path)

3) def destroy
    if @professional_development.destroy
      redirect_to professional_developments_url, notice: "Professional development deleted successfully."
    else
      redirect_to professional_developments_url, alert: "Failed to delete professional development." --> **UNNECESSARY if/else clause** 

There is no reason why deleting a professional development should fail. In all other places in the app where stuff is destroyed there is no error checking to see if deleting something fails. So remove the if/else stuff.

4) At the bottom of app/views/professional_developments/_form.html.erb, I noticed that you basically just copy-pasted the script from one of the other views. Is there a way that we can DRY out this code and maybe put the script in some third location where it can be loaded in for both the views?

5) def registration_open
    @professional_development.registration_open ? "Yes" : "No"

Is this function even necessary? It doesn't seem to be used anywhere in the code...

**Comment by ArushC** [comment]
Also nooo.... there's a merge conflict :(

File: app/controllers/pd_registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class PdRegistrationsController < ApplicationController
+  before_action :require_login
+  before_action :require_admin
+  before_action :set_pd_registration, only: [:show, :edit, :update, :destroy]
+  before_action :set_professional_development, only: [:new, :create, :edit, :update, :destroy]
+
+  def index
+    @pd_registrations = PdRegistration.where(professional_development_id: @professional_development.id)
```
Comment (fp456) [inline app/controllers/pd_registrations_controller.rb:10]: Line 10 not tested

File: app/controllers/pd_registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      render "professional_developments/show"
+    end
+  end
+
+  def destroy
+    if @pd_registration.destroy
+      redirect_to professional_development_path(@professional_development),
+                  notice: "Registration was successfully cancelled."
+    else
+      flash.now[:alert] = @pd_registration.errors.full_messages.to_sentence
```
Comment (fp456) [inline app/controllers/pd_registrations_controller.rb:51]: This path was not tested

File: app/controllers/pd_registrations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      render "professional_developments/show"
+    end
+  end
+
+  private
+  def set_pd_registration
+    @pd_registration = PdRegistration.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
+    redirect_to professional_development_path, alert: "Registration not found."
```
Comment (fp456) [inline app/controllers/pd_registrations_controller.rb:60]: if possible, test

File: app/controllers/pd_registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def set_professional_development
+    @professional_development = ProfessionalDevelopment.find_by(id: params[:professional_development_id])
+    unless @professional_development
+      redirect_to professional_developments_path, alert: "Professional Development not found."
```
Comment (fp456) [inline app/controllers/pd_registrations_controller.rb:66]: missing test

File: app/controllers/professional_developments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    else
+      redirect_to professional_developments_url, alert: "Failed to delete professional development."
+    end
+  end
+
+  private
+  def set_professional_development
+    @professional_development = ProfessionalDevelopment.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
+    redirect_to professional_developments_url, alert: "Professional development not found."
```
Comment (fp456) [inline app/controllers/professional_developments_controller.rb:54]: test if possible

**Review by fp456 — COMMENTED** [review]
Looks good so far! Just some extra testing needed, which is necessary for a large PR like this. Goood job! 🙌

**Comment by ArushC** [comment]
@perryzjc what's the progress on this PR. Is it ready to be merged yet?
